main.tf

```diff
@@ -67,11 +75,24 @@ resource "aws_subnet" "public" {
   tags = "${merge(var.tags, var.public_subnet_tags, map("Name", format("%s-public-%s", var.name, element(var.azs, count.index))))}"
 }
 
+resource "aws_subnet" "public_ipv6" {
```

Comment:
Making `aws_subnet` resources conditionally is a bad idea, because it require twice more code for some things. Instead, we should always use the same `aws_subnet` resource (`public` and `private`). I think it should be possible, because ipv6-related features are the same as for ipv4 (with the exception of `aws_egress_only_internet_gateway` which is only for ipv6).

Follow up questions:
1. Will `ipv6_cidr_block` work with empty value?
1. What will the value of `aws_vpc.this.ipv6_cidr_block` be when VPC has not enabled IPv6?
1. Why did you specify the number of bits as "8" as an argument of `cidrsubnet`?

Reply:
Hi Anton,
I agree that making ```aws_subnet``` resources conditionally is not ideal - if you had a chance to review the discussion in the related issue you'll see I originally went down the path of using the same ```private``` and ```public``` subnet resources but ran into some issues due to the ```ipv6_cidr_block``` behavior.

To address your questions directly:
1.  ```ipv6_cidr_block``` subnet argument will work with an empty value, but the following 
```ipv6_cidr_block         = "${var.enable_ipv6 ? cidrsubnet(aws_vpc.this.ipv6_cidr_block,8,count.index) : ""}"```
won't work due to the answer to 2. below 
2.  ```aws_vpc.this.ipv6_cidr_block``` is undefined when the ```assign_generated_ipv6_cidr_block``` argument is ```false``` in vpc resource definition, which resulted in the conditional referenced in 1. failing with the following error: 
```* module.vpc.aws_subnet.public[0]: Resource 'aws_vpc.this' does not have attribute 'ipv6_cidr_block' for variable 'aws_vpc.this.ipv6_cidr_block'```
3.  I specified 8 bits in ```cidrsubnet``` because AWS ipv6 blocks are autoassigned a /56 mask and the AWS ipv6 subnets require a /64 mask, so this should always be 8 (i think)

Reply:
Thanks for the clarification. I want to try the code in this PR and see if I can make something around 1 and 2 during Monday evening.

Reply:
@antonbabenko looks like there was a recently submitted [PR](https://github.com/terraform-providers/terraform-provider-aws/pull/2103) that would fix the [issue](https://github.com/terraform-providers/terraform-provider-aws/issues/688) we are seeing.

If we wait until this is fixed we won't need to create conditional subnet resources.  Additionally, it looks like there is a "fix" coming in 0.11.x (see 0.11.0-beta1 [changelog](https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md#0110-beta1-november-3-2017)) that may cause this PR code to break because of the new ```output``` validation (I assume that if our conditionally created subnet resources are not created, terraform would throw an error due to the exception in the ipv6_* output interpolations).

I'm thinking we can sit on this until that PR is merged and I can refactor the ipv6 support into the existing the resources, and hopefully at the same time provide forward compatibility.  Let me know your thoughts :)

Reply:
Great finding! Absolutely, let's wait for this PR. Terraform AWS provider is the best place to fix this kind of issues.